tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,192 @@
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
+def convert_arguments(args: List[Argument]) -> Tuple[List[Expr], List[str]]:
+    argument_str = "c10::IValue {arg_name} = std::move(peek(stack, {pos}, {args_num}));"
+    expr_list = []
+    pop_ivalue = []
+    code_list = []
+    for i, arg in enumerate(args):
+        pop_ivalue.append(argument_str.format(arg_name=arg.name, pos=i, args_num=len(args)))
+        expr, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
+    pop_ivalue.append("")
+    return expr_list, pop_ivalue + code_list
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[Expr, List[str]]:
+    nctype = argumenttype_type(t=t, mutable=mutable, binds=arg_name)
+    ctype = nctype.cpp_type(strip_ref=True)
+
+    if isinstance(t, BaseType):
+        out_name = f"{arg_name}_base"
+        code = [f"{ctype} {arg_name}_base = {arg_name}.to<{ctype}>();"]
+    elif isinstance(t, OptionalType):
+        out_name = arg_name + "_opt_out"
+        code = _gen_code_optional_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    elif isinstance(t, ListType):
+        out_name = arg_name + "_list_out"
+        code = _gen_code_list_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    else:
+        raise Exception(f"Cannot handle type {t}. arg_name: {arg_name}")
+    expr = Expr(expr=out_name, type=nctype)
```

Comment:
Oh I see why you had to add the extra rule to `translate()` - right here, you're manually creating the `Expr` object.

You could manually write logic here like `Expr(expr=out_name if out_name != "memory_format" else SpecialArgName.possibly_redundant_memory_format, ...)`

...But I don't think you should actually be manually creating the `Expr` object at all (which will end up being fragile, since you have to create it exactly the same way that the C++ api codegen does it)

Instead, you should rely on the C++ api codegen to do it for you (which you can see happens [here](https://github.com/pytorch/pytorch/blob/09e54ffec3d01d2742b7bad5b5cbd5ffc56bcf33/tools/codegen/api/cpp.py#L258)). We already have a way for generating those bindings, through the `CppSignature` object. (you can see the functionalization codegen doing something similar [here](https://github.com/pytorch/pytorch/blob/63429bf4b3f491164f9896701134d81162b27469/tools/codegen/gen_functionalization_type.py#L52))

One way would be to do something like:
```
# pass this CppSignature(...).most_faithful_signature()
def convert_arguments(sig: CppSignature) -> Tuple[List[Binding], List[str]]:
    code_list = []
    binding_list = []
    args = sig.arguments()
    for i, a in enumerate(args):
        # get the code for unwrapping the ivalue, and the name of the variable we unwrapped
        unboxed_arg_name, code = argumenttype_ivalue_convert(a.argument.type, a.name, mutable=a.argument.is_write)
        code_list.append(f"""
c10::IValue {a.name} = std::move(peek(stack, {i}, {len(args)}));
{'\n'.join(block for block in code)}
        # add the binding for our new unwrapped variable to the context
        binding_list.append(a.with_name(unboxed_arg_name)
        ...
```